sbi/src/attestation.rs

```diff
@@ -5,29 +5,147 @@
 use crate::error::*;
 use crate::function::*;
 
+use flagset::{flags, FlagSet};
+
 /// The data blob passed to the GetEvidence call must be 64 bytes long.
 pub const EVIDENCE_DATA_BLOB_SIZE: usize = 64;
 
-/// Attestation evidence formats.
-#[derive(Copy, Clone, Debug)]
-#[repr(u64)]
-pub enum EvidenceFormat {
-    /// Single layer DICE TCB
-    /// https://trustedcomputinggroup.org/resource/dice-attestation-architecture/
-    DiceTcbInfo = 0,
-
-    /// X.509 extension for multiple layers DICE TCB.
-    /// https://trustedcomputinggroup.org/resource/dice-attestation-architecture/
-    DiceMultiTcbInfo = 1,
-
-    /// Open DICE profile
-    /// https://pigweed.googlesource.com/open-dice/+/HEAD/docs/specification.md
-    OpenDice = 2,
+/// The maximum number of measurement registers in the attestation capabilities.
+pub const MAX_MEASUREMENT_REGISTERS: usize = 32;
+
+flags! {
+    /// Attestation evidence formats.
+    #[derive(Default)]
+    #[repr(u8)]
+    pub enum EvidenceFormat: u8 {
+        /// Single layer DICE TCB
+        /// https://trustedcomputinggroup.org/resource/dice-attestation-architecture/
+        #[default]
+        DiceTcbInfo = 1,
+        /// X.509 extension for multiple layers DICE TCB.
+        /// https://trustedcomputinggroup.org/resource/dice-attestation-architecture/
+        DiceMultiTcbInfo = 2,
+        /// Open DICE profile
+        /// https://pigweed.googlesource.com/open-dice/+/HEAD/docs/specification.md
+        OpenDice = 4,
+    }
+}
+
+/// A list of supported hash algorithms.
+#[derive(Copy, Clone, Default, Debug)]
+#[repr(u8)]
+pub enum HashAlgorithm {
+    /// SHA-384
+    #[default]
+    Sha384 = 1,
+    /// SHA-512
+    Sha512 = 2,
+}
+
+/// Attestation Capabilities
+///
+/// This structure exposes the supported attestation capabilities to the SBI
+/// GetCapabilities caller. It lets the caller know which hash algorithms,
+/// evidence formats, and measurements mappings the SBI implementation supports.
+#[repr(C)]
+#[derive(Default)]
+pub struct AttestationCapabilities {
+    /// The TCB Secure Version Number.
+    pub tcb_svn: u64,
+    /// The supported hash algorithm.
+    pub hash_algorithm: HashAlgorithm,
+    /// The supported evidence formats. This is a bitmap.
+    pub evidence_formats: FlagSet<EvidenceFormat>,
+    /// Number of static measurement registers.
+    pub static_measurements: u8,
+    /// Number of runtime measurement registers.
+    pub runtime_measurements: u8,
+    /// Array of all measurement register descriptors.
+    pub measurement_registers: [MeasurementRegisterDescriptor; MAX_MEASUREMENT_REGISTERS],
+
+    // Track the current measurement_registers array index.
+    measurement_registers_idx: usize,
```

Comment:
Is this only used when building the `AttestationCapabilities` struct, or does the receiver of this struct also use it to determine how many `measurement_registers` are populated? If it's the former, then I'm not sure it belongs in the ABI.

Reply:
I think you're right. I've added an index to the `add_measurement_register` call and let the SBI implementation decide where the descriptor should land in the array.